draft-thomson-http-binary-message.md

```diff
@@ -29,21 +29,594 @@ This document defines a binary format for representing HTTP messages.
+
+Other values cause the message to be invalid; see {{invalid}}.
+
+
+## Request Control Data
+
+The control data for a request message includes four values that correspond to
+the values of the `:method`, `:scheme`, `:authority`, and `:path` pseudo-header
+fields described in HTTP/2 (Section 8.1.2.3 of {{!H2}}). These fields are
+encoded, each with a length prefix, in the order listed.
+
+The rules in Section 8.1.2.3 of {{!H2}} for constructing pseudo-header fields
+apply to the construction of these values. However, where tge `:authority`
+pseudo-header field might be omitted in HTTP/2, a zero-length value is encoded
+instead.
+
+The format of request control data is shown in {{format-request-control-data}}.
+
+~~~ quic-format
+Request Control Data {
+  Method Length (i),
+  Method (..),
+  Scheme Length (i),
+  Scheme (..),
+  Authority Length (i),
+  Authority (..),
+  Path Length (i),
+  Path (..),
+}
+~~~
+{: #format-request-control-data title="Format of Request Control Data"}
+
+
+## Response Control Data
+
+The control data for a request message includes a single field that corresponds
+to the `:status` pseudo-header field in HTTP/2 {{!H2}}. This field is encoded
+as a single variable length integer, not a decimal string.
+
+The format of final response control data is shown in
+{{format-response-control-data}}.
+
+~~~ quic-format
+Final Response Control Data {
+  Status Code (i) = 200..599,
+}
+~~~
+{: #format-response-control-data title="Format of Final Response Control Data"}
+
+
+### Informational Status Codes {#informational}
+
+This format supports informational status codes (see Section 15.2 of
+{{!HTTP}}). Responses that include information status codes are encoded by
+repeating the response control data and associated header section until the
+final status code is encoded.
+
+The format of the informational response control data is shown in
+{{format-informational}}.
+
+~~~ quic-format
+Informational Response Control Data {
+  Status Code (i) = 100..199,
+}
+~~~
+{: #format-informational title="Format of Informational Response Control Data"}
+
+A response message can include any number of informational responses. If the
+response control data includes an informational status code (that is, a value
+between 100 and 199 inclusive), the control data is followed by a header
+section (encoded with known- or indeterminate- length according to the framing
+indicator). After the header section, another response control data block
+follows.
+
+
+## Header and Trailer Field Lines {#fields}
+
+Header and trailer sections consist of zero or more field lines; see Section 5
+of {{!HTTP}}. The format of a field section depends on whether the message is
+known- or intermediate-length.
+
+Each field line includes a name and a value. Both the name and value are
+non-zero length sequences of bytes. The format of a field line is shown in
+{{format-field-line}}.
+
+~~~ quic-format
+Field Line {
```

Comment:
Do we see a way forward to have a binary encoding of structured fields? While we don't need to specify that here, it'd be nice to leave it open.

Reply:
Does anything need to change to accommodate them here? I figured they'd just be values with a different encoding.

Reply:
A new format is what I was thinking, or maybe use the next bit in the framing indicator.  The latter has some risks though.